63 test results

cwmeijer opened the pull request:
This PR adds an assert statement to every existing test aimed at experiments. The assert tests the result, for instance, the loss, or the rank. I didn't discuss this kind of assertion with anyone, so it's definitely worth looking at these specifics assertions and see if you guys agree with what I did. I also had to change the experiments and scripts in order to be able to read the results from the tests.

cwmeijer committed:
add asserts for results in experiment tests (refs #63)

cwmeijer commented:
I'll have a look at the conflicts now, so I'll turn this into a draft again.

cwmeijer committed:
Merge branch 'master' into 63-test-results

cwmeijer committed:
Merge branch 'master' into 63-test-results

bhigy commented on platalea/basic.py:
```diff
@@ -170,6 +171,9 @@ def val_loss(net):
             wandb.log(result)
 
+    # Return loss of the final model for automated testing
+    return {'final_loss': loss_value}
```
Is there a reason why we return different criteria (`validation_loss`, `final_loss`) for asr and basic experiments? I would make this consistent by returning the same criteria or both.

bhigy replied:
Or maybe better, do as in (e.g.) `mtl.py` and save and return all intermediate scores.

cwmeijer replied:
I'll have a look if it can always return the result dict (that is written to json anyway).

bhigy replied:
You might have to save the results at each epoch.

Little reminder: please also check `experiments/flickr8k/pip_ind.py`.

cwmeijer committed:
fix tests while maintaining sensible result returns
The challenge here is to have the experiments return sensible results for the
user while also meaningfull to test against (sensitive to code/logic changes).
This commit manages to do this for most tests. refs #63

cwmeijer committed:
add meaningful test results to transformer and basic tests (refs #63)

cwmeijer committed:
add approximate checking in experiment tests (refs #63)

cwmeijer committed:
add pandas to test environment (refs #63)

bhigy reviewed (commented):
Looks pretty good to me but I am wondering whether it makes sense to have the `step_loss` in the results. That is only the loss for one training step right? The rest of the results are computed on the full validation set, which makes it a bit confusing to me.

cwmeijer commented:
All experiment functions now return sensible results that can be useful to the user. These consist of performance metrics of every step. Because the tests are usually performing only a single update step, with minimal input data, the performance metrics are often trivial. For instance, rank.10 is always 1 because, in the tests, there are no more instances than 10. I, therefore, added a training loss to the results to have the results include a measure very sensitive to logic/code changes. Because different machines resulted in slightly different roundings, I had to use an approximate checker instead of an exact one. Therefore, I chose to include pandas in the test environment.

cwmeijer committed:
Merge branch 'master' into 63-test-results

cwmeijer commented:
The step_loss is the training loss at that time step. So, just another performance metric about the current time step. Of course, it is not the most useful performance metric ever, but it is somewhat informative for a user, and of course useful for the test. If you don't agree, we can look for other solutions.

bhigy commented:
No, I agree. I didn't really think about the degenerate scores we get when testing with a just one batch. With this, it makes sense to add the last training loss.

Is there a specific reason why you don't add it in `asr.py` though?

bhigy commented:
Checking the code again, I think `step_loss` is actually a bit misleading. This is actually the mean training loss over the whole epoch right?

cwmeijer committed:
rename step_loss, added average_loss to all experiments (ref #63)
